posts/2024-05-07-ultimate-string-guide.md

```diff
@@ -0,0 +1,625 @@
+
+Here we see a different technique than surrogates. UTF-8 uses magic bits in the first byte to signal how many bytes in total
+must be read for translating to a code point.
+
+Notable properties of UTF-8 are:
+
+- it is ASCII backwards compatible: a program written for UTF-8 will also understand plain ASCII encoding
+- unicode code points in the surrogate range `U+D800` to `U+DFFF` are considered invalid byte sequences
+  - as a result: UTF-8 only expresses Unicode Scalar Values
+
+### Unicode summary
+
+Given the above encodings, let's have another look at our table from above:
+
+| character          | code point       | Hex UTF-8 | Hex UTF-16 | Hex UTF-32 |
+|--------------------|------------------|-----------|------------|------------|
+| a                  | U+0061           | 61        | 0061       | 00000061   |
+| b                  | U+0062           | 62        | 0062       | 00000062   |
+| 쟬                 | U+C7EC           | ec 9f ac  | c7ec       | 0000c7ec   |
+| 🇯                 | U+1F1EF          | f0 9f 87 af | d83c ddef | 0001f1ef  |
+| 🇵                 | U+1F1F5          | f0 9f 87 b5 | d83c ddf5 | 0001f1f5  |
+| 🇯🇵               | U+1F1EF, U+1F1F5 | f0 9f 87 af, f0 9f 87 b5 | d83c ddef, d83c ddf5 | 0001f1ef, 0001f1f5 |
+
+The interested reader is welcome to verify those values (at least for UTF-8 and UTF-16).
+
+We now understand:
+
+- the character encoding is the mapping of code points to visible characters
+- UTF-8, UTF-16 and UTF-32 are text encodings with different trade offs
+- surrogates are a special case for UTF-16 (`Unicode Scalar Values = Unicode Code Points - surrotages`)
+
+Going back to the definition of "character", we now see the confusion:
+
+- a surrogate can hardly be a visible character
+- the visible character  🇯🇵 needs two code points to be expressed (and there are many others)
+
+This has lead to yet another definition: **"Grapheme Cluster"**. This is specified by the [Unicode Standard Annex #29](https://www.unicode.org/reports/tr29/), which deals with determining boundaries between characters, words and sentences.
+It is, again, quite technical, but is much closer to "user visible character".
+
+## Back to Haskell String type
+
+Now that we know what a **Unicode Code Point** is, we also understand that the Haskell String type
+has essentially no *text encoding*. It is just a linked list of those code points (a subset of `Int`, in fact).
+This can be a nice property, e.g. as an intermediate representation when converting between encodings (say UTF-8 to UTF-16).
+
+However, it is a **questionable default for a String type**, because:
+
+- it is inefficient for large text (carries the overhead of a linked list with thunks for every `Char`)
```

Comment:
It might be useful to copy a diagram from page 2 of https://github.com/Bodigrim/my-talks/blob/master/zurihac2022/slides.pdf, demonstrating how wasteful `String` is.